### PR TITLE
Started OSL layered closures support.

### PR DESCRIPTION
**Pull request by est77**
Changes needed to support future layered OSL closures.

### CODE REVIEW

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 1]
```diff
@@ -927,6 +941,27 @@ namespace
             shading_system.register_closure(name(), id(), params, 0, 0);
         }
     };
+
+    //
+    // Layered Closures.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: *closures*

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 2]
```diff
@@ -927,6 +941,27 @@ namespace
...
+        void* substrate;
+    };
+
+    const OSL::ClosureColor* get_nested_closure_color(
+        const size_t    closure_id,
+        const void*     params)
+    {
+        assert(closure_id >= FirstLayeredClosure);
+
+        const LayeredClosureBaseParams *p =
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Move star to the type's side.

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 3]
```diff
@@ -1332,6 +1375,13 @@ void CompositeEmissionClosure::process_closure_tree(
 
             if (c->id == EmissionID)
                 m_total_weight += weight * Color3f(c->w);
+            else if(c->id >= FirstLayeredClosure)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Insert space after `if`.

File: src/appleseed/renderer/kernel/shading/closures.cpp [hunk 4]
```diff
@@ -1374,7 +1424,17 @@ namespace
               default:
                 {
                     const OSL::ClosureComponent* c = reinterpret_cast<const OSL::ClosureComponent*>(closure);
-                    return c->id == closure_id ? Color3f(c->w) : Color3f(0.0f);
+
+                    if (c->id == closure_id)
+                        return Color3f(c->w);
+                    else if(c->id >= FirstLayeredClosure)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.cpp]: Same here.

File: src/appleseed/renderer/kernel/shading/closures.h [hunk 5]
```diff
@@ -121,6 +121,9 @@ enum ClosureID
     HoldoutID,
     TransparentID,
 
+    // Layered BSDF closures. (Must be last.)
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/closures.h]: It's not really last, isn't it? :)

File: src/appleseed/renderer/kernel/shading/oslshadergroupexec.cpp [hunk 6]
```diff
@@ -51,7 +51,10 @@ OSLShaderGroupExec::OSLShaderGroupExec(OSL::ShadingSystem& shading_system)
   : m_osl_shading_system(shading_system)
   , m_osl_thread_info(shading_system.create_thread_info())
   , m_osl_shading_context(shading_system.get_context(m_osl_thread_info))
+  , m_osl_mem_used(0)
 {
+    m_osl_mem_pool = new char[OSLMemPoolSize + OSLMemAlignment];
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/oslshadergroupexec.cpp]: If the size is known at compile-time, can't this be statically allocated in the class?

File: src/appleseed/renderer/kernel/shading/oslshadergroupexec.cpp [hunk 7]
```diff
@@ -231,6 +236,23 @@ void OSLShaderGroupExec::do_execute(
...
+
+    reset_osl_mem_pool();
+}
+
+void* OSLShaderGroupExec::osl_mem_alloc(const size_t size) const
+{
+    if APPLESEED_UNLIKELY(m_osl_mem_used + size >= OSLMemPoolSize)
+        return 0;
+
+    char *ptr = m_osl_mem_pool_start + m_osl_mem_used;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/oslshadergroupexec.cpp]: Move star.

File: src/appleseed/renderer/kernel/shading/oslshadergroupexec.cpp (hunk 6 above)
Comment (est77) [inline src/appleseed/renderer/kernel/shading/oslshadergroupexec.cpp]: In the stack?

Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/oslshadergroupexec.cpp]: I see that the size if 128 KB. So maybe not :)

Comment (est77) [inline src/appleseed/renderer/kernel/shading/oslshadergroupexec.cpp]: This is allocated only once per thread, so I think it's ok.
We could use this memory for other things in the future too.

**Comment by est77** [comment]
Fixed & rebased

**Pull request merged**